Fix area tiling issue with fx files in area_statistics, added calculate_grid option, added several unit tests to area statistics, clarified several exceptions in area_statistics

ledm opened the pull request:
Before you start, please read [CONTRIBUTING.md](https://github.com/ESMValGroup/ESMValCore/blob/master/CONTRIBUTING.md).

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [ ] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [ ] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [ ] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #445

ledm commented:
Question to the reviewers. At the moment, the tests are failing because of the absence of FX files. How do I write a test that provides an `fx_files`?

Should we add an small sample areacello fx file somewhere to the repository?

valeriupredoi commented:
no need to, you can create a wee cube and save it with a patched func like in `tests/integration/test_recipe.py` - I can do that for you in this test on Monday :beer:

ledm commented:
Making some progress on this, and I've added several sensible tests (with assert raises) for the area_statistics calculation.

ledm commented:
I've changed the implementation of `area_statistics` so that if you want to use iris to calculate the grid cell area, you need to explicitly ask for this with the option `calculate_grid`. 

To make sure the functionaility of this preprocessor is explicitly clear, I've added a few failing clauses: 
- If `calculate_grid` is set to `True, an exception will be raised if your grid is irregular (latitide has more than 1 dimensions). This is because iris can't calculate the cell area for complex/irregular grids.

- If `calculate_grid` is set to `True, an exception will be raised if you also provide FX files in your recipe. You can't provide both an fx file and attempt to calculate the grid. 

- These tests only occur if the operator requires the grid definitions (many do not).

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -177,13 +177,18 @@ def tile_grid_areas(cube, fx_files):
             elif cube.ndim == 3 and grid_areas.ndim == 2:
-                grid_areas = da.tile(grid_areas, [cube.shape[0], 1, 1])
+                tiled = da.array([grid_areas for i in range(cube.shape[0])])
+                grid_areas = da.stack(tiled, axis=0).compute()
```
replace the two `elif`'s with:
```
elif (cube.ndim, grid_areas.ndim) in [(4, 3), (3, 2)]:
```
and write the defs for `tiled` and `grid_areas` only once

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -243,33 +255,36 @@ def area_statistics(cube, operator, fx_files=None):
-    if grid_areas is None or not grid_areas.any():
-        cube = guess_bounds(cube, coord_names)
+    if not calculate_grid:
+        grid_areas = tile_grid_areas(cube, fx_files)
```
is `fx_files` an optional argument initialized with `None` for `tile_grid_areas()`?

ledm replied:
This code is outdated now...

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -243,33 +255,36 @@ def area_statistics(cube, operator, fx_files=None):
+    if not calculate_grid:
+        grid_areas = tile_grid_areas(cube, fx_files)
+    else:
+        cube = guess_bounds(cube, ['longitude', 'latitude'])
```
check if previous bounds exist

ledm replied:
`guess_bounds` does this check already.

valeriupredoi replied:
no it don't, it fails if there are previous bounds

ledm replied:
```
# guess bounds tool
def guess_bounds(cube, coords):
    """Guess bounds of a cube, or not."""
    # check for bounds just in case
    for coord in coords:
        if not cube.coord(coord).has_bounds():
            cube.coord(coord).guess_bounds()
    return cube
```

valeriupredoi replied:
ah yes, that's the wrapper of guess bounds from iris, my bad (I put the wrapper in and I forget about it)

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -177,22 +177,47 @@ def tile_grid_areas(cube, fx_files):
+        grid_areas = da.stack(tiled, axis=0).compute()
+    elif cube.ndim == 3 and grid_areas.ndim == 2:
+        tiled = da.array([grid_areas for i in range(cube.shape[0])])
+        grid_areas = da.stack(tiled, axis=0).compute()
```
new code but still the same repetition of lines, man, pls use the suggestion from my previous comment

ledm commented on tests/unit/preprocessor/_area/test_area.py:
```diff
@@ -15,6 +15,83 @@
+    gridir.add_aux_coord(lon_coord, data_dims=(0, 1))
+    fn = 'fx_test_areacello.nc'
+    iris.save(gridir, fn)
+    return fn
```
@valeriupredoi I wrote some tests (`test_area_statistics_mean_ir` and test_area_statistics_sum_ir`) and they are failing because IRIS doesn't like the fx cube I'm making here:

Any thoughts?

valeriupredoi replied:
line 3277 from `cube.py` is trying to view (memory map) your **numpy** array but your weights is a dask array that doesn't have `view` as attribute:
```
In [42]: b                                                                                                                                        
Out[42]: dask.array<array, shape=(2,), dtype=int64, chunksize=(2,), chunktype=numpy.ndarray>

In [43]: b.view()                                                                                                                                 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-43-4a91f95886bf> in <module>
----> 1 b.view()

TypeError: view() missing 1 required positional argument: 'dtype'
```
you need to realize your weights before calling the collapse (I don't think they've implemented lazy weights yet in iris)

valeriupredoi replied:
just replace:
```
@@ -322,7 +322,7 @@ def area_statistics(cube, operator, fx_files=None, calculate_grid=False):
                          '({})'.format(cube.shape, grid_areas.shape))
 
     return cube.collapsed(['longitude', 'latitude'], operation,
-                          weights=grid_areas)
+                          weights=np.array(grid_areas))
```
note that the two tests will fail again but this time the result is not the same as expected (first is 1.5 when you check on 1.0 and the second is 7.381511e+12 when you check on 1.0) - I suspect you are not adding the right results spat by area statistics

ledm replied:
I did something like this, but used:
```
    if isinstance(grid_areas, da.Array):
        grid_areas = grid_areas.compute()
```

valeriupredoi replied:
ya that'll do! bit confused as to why your area statistic is such a monster number (7e12)

valeriupredoi commented:
can you pls give it a meaningful title?

ledm commented:
> can you pls give it a meaningful title?
Done

valeriupredoi commented:
> > can you pls give it a meaningful title?
> > Done

Almost :grin: All good, man! Can you pls tell me why you are expecting such a monster number from your test result (order 7e12)?

ledm commented:
yes, it's a weighted sum and the weights are real cell areas for 5x5 degree grid cells. It's a big number.

valeriupredoi commented:
> yes, it's a weighted sum and the weights are real cell areas for 5x5 degree grid cells. It's a big number.

ok so if you are expecting it to be this of a biggie, all good from me buds

ledm commented:
@schlunma @zklaus any chance that you could take a look at this please!

ledm commented:
It's worth mentionning that this PR changes the interface to the `area_statistics` preprocessor (in a sensible direction). However, some recipes may need to be changed as the `area_statistics` operators `mean` and `sum` will now not work without either a `fx_files` or a `calculate_grid: True` option. 

I guess that this means that we need to make a new ESMValTool branch and change a couple recipes.

valeriupredoi commented:
Can you pls do that on ESMValTool so  @mattiarighi can test the coupled core/tool modules? Code-wise this is all ok by me :beer:

schlunma commented:
What about cases where some data sets provide fx files and others don't (e.g. `OBS` ones)? With these changes it's not possible to calculate the areas in case there are no fx files, right? Maybe we change the name from `calculate_grid` to `allow_grid_calculation` and allow the use of `allow_grid_calculation` together with `fx_files`?

Otherwise I'm okay with the changes :+1:

ledm commented:
> What about cases where some data sets provide fx files and others don't (e.g. `OBS` ones)? With these changes it's not possible to calculate the areas in case there are no fx files, right? Maybe we change the name from `calculate_grid` to `allow_grid_calculation` and allow the use of `allow_grid_calculation` together with `fx_files`?
> 
> Otherwise I'm okay with the changes 👍

In that case, I've been assuming that we just apply a different preprocessor to the obs datasets. Is that sensible or are there cases where we need obs and models to use the same preprocessor chain?

schlunma commented:
That's of course always possible, but I think this is really not user-friendly, since it implies adding a new entry to the `variable` section of a diagnostic and adding `additional_datasets` at the correct places. Not hard if you are used to that, but might be problematic for new users.

ledm commented:
I don't like the idea of having that level of (hidden) flexibility. If people want to calculate their own grid areas, they should explicitly ask for that. 

Just so that everyone is aware that that is whats happening and that that is what they want.

schlunma commented:
But if the user has to explicitly specify `allow_grid_calculation` the feature is not hidden, isn't it? If that is not desired, only `fx_files` should be used, which gives an error if no files are available. If only `allow_grid_calculation` is given, no `fx_files` are used. Fairly transparent in my opinion.

ledm commented:
So, what are we waiting for on this PR?

mattiarighi commented:
To be reviewed by @bouweandela and tested, as usual.

valeriupredoi commented:
> But if the user has to explicitly specify `allow_grid_calculation` the feature is not hidden, isn't it? If that is not desired, only `fx_files` should be used, which gives an error if no files are available. If only `allow_grid_calculation` is given, no `fx_files` are used. Fairly transparent in my opinion.

@ledm have you addressed in any way @schlunma comment above? If no code change is needed then pls give a motivation that might be good to put in the documentation as well :beer: Anyways, I *think* this PR will only work after #439 has been merged no? But by all means let's hurry up and get both this and 439 reviewed pls

bouweandela commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -243,33 +272,60 @@ def area_statistics(cube, operator, fx_files=None):
     if not fx_files and cube.coord('latitude').points.ndim == 2:
         logger.error(
-            'fx_file needed to calculate grid cell area for irregular grids.')
+            'fx_file needed for weighted calcuions on irregular grids.')
```
```suggestion
            'fx_file needed for weighted calculations on irregular grids.')
```

bouweandela commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -243,33 +272,60 @@ def area_statistics(cube, operator, fx_files=None):
-    # TODO: implement weighted stdev, median, s var when available in iris.
-    # See iris issue: https://github.com/SciTools/iris/issues/3208
+    if isinstance(grid_areas, da.Array):
+        grid_areas = grid_areas.compute()
```
Why are you realizing the data? Is it not possible to use a dask array as a weight?

ledm replied:
Nope, iris only takes numpy arrays as weights.

bouweandela commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -177,22 +177,44 @@ def tile_grid_areas(cube, fx_files):
 
 # get the area average
-def area_statistics(cube, operator, fx_files=None):
+def area_statistics(cube, operator, fx_files=None, calculate_grid=False):
```
I think it would be nicer to define this function as
```python
def area_statistics(cube, operator, weights=None):
```
and `weights` can be either the path to an fx file, the string `'calculate'` or `None` for unweighted statistics.

bouweandela commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -177,22 +177,44 @@ def tile_grid_areas(cube, fx_files):
+    iris.cube.Cube
+        Freshly tiled grid areas cube.
+    """
+    if cube.ndim == grid_areas.ndim:
```
I think this entire function could be replaced by
```python
grid_areas = np.broadcast_to(grid_areas, cube.shape[:-grid_areas.ndim] + grid_areas.shape)
```

bouweandela replied:
or even:
```python
grid_areas = np.broadcast_to(grid_areas, cube.shape)
```
or am I missing something?

bouweandela commented:
> That's of course always possible, but I think this is really not user-friendly, since it implies adding a new entry to the `variable` section of a diagnostic and adding `additional_datasets` at the correct places. Not hard if you are used to that, but might be problematic for new users.

If this is problematic for new users, I think we need to improve our documentation. I agree with @ledm that it should be no problem to use different preprocessors for different datasets.

bouweandela commented:
> using only fx files, using only calculated grid areas or using both with a preference on using fx files

I think that would also be fine, so long as it doesn't silently fall back to calculated areas when you specify you want to use an fx file.

The reason why land/sea masks are different, at least intuitively, is that you can really easily see the difference in masking when you plot the data, while with statistics you'll only get some numbers out, so it's less easy to see if things go wrong. Maybe it would be good to make the land/sea mask preprocessor function more explicit in whether or not you want natural earth as a fallback.

sloosvel commented:
Sorry but I've been waiting quite some time for the area statistics to be fixed. I do not mind doing workarounds but now we have users wanting to use the preprocessor on irregular grids, and for some reason the code in master fails at `da.tile`. If this is not ready, could at least the code on master go back to using `np.tile` ?

zklaus commented:
@ledm I'm afraid this is on your desk. I particularly agree with @bouweandela that `tile` is probably uncalled for. Let me know if you would like to discuss the broadcast approach.

ledm commented:
What are the differences between `np.tile`, `da.tile` and `np.broadcast_to`. As far as I can tell, from the users perspective, they are all just different ways to produce the same output. I'm not wise enough to know which one is optimised for our use case. 

Note that this entire problem was caused by `np.tile` and `da.tile` having a different interface, so they are not perfectly compatible.

Is there a dask version of `broadcast_to` to consider as well?

Having said that, functionality should really be included in iris, when cell methods are built into the data cube.

bouweandela commented:
> Note that this entire problem was caused by np.tile and da.tile having a different interface, so they are not perfectly compatible.

I think this may have been solved since dask version 2.8.0 (see [here](https://docs.dask.org/en/latest/changelog.html#id128)).

zklaus commented:
@ledm, tiling creates a new array with copies of data, broadcasting creates a new [view](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.view.html) using the same data memory with only a bit of logic to make the indices point to the right place.

As a consequence, broadcasting works in constant time and memory (O(1)), tiling needs memory and time that depends on the shape of the target, hence is much more expensive. Always prefer broadcasting, if possible! Only when you actually need the new array to have and manipulate further consider using tiling. (Cf the numpy documentation on tiling:
> Note: Although tile may be used for broadcasting, it is strongly recommended to use numpy’s broadcasting operations and functions.

)


Ok, that's the two concepts, now how can we use them? Tiling using `np.tile`/`da.tile` seems clear enough. Broadcasting is a bit more comprehensive, but in many cases simpler. Broadcasting always happens when combining two arrays of different shape with operations that would work on same-shape arrays. Examples:
```python
a = np.random.randn(3, 3)  # create random 3x3 matrix
3. * a  # multiply scalar (shape: (,)) with matrix (shape: (3, 3)); scalar will be broadcast to shape (3, 3)
b = np.array([1, 2, 3])  # create 3 vector
a * b  # multiply `a` (shape: (3, 3)) with b (shape: (3,)); b will be broadcast to (3, 3)
a * b[np.newaxis, :]  # the same as the previous line
a * b[:, np.newaxis]  # almost the same, but interpret b as column vector instead of a row vector
```
So, you likely have been broadcasting all the time, for a long time, without knowing it :smile:.

Sometimes, numpy/dask needs a bit of help to do the broadcasting correctly. The simplest way is when you need to specify where to insert axes to get one array to the shape of the other; this you can see in the examples above with `np.newaxis`. Sometimes even that is not enough, for example when you already know the shape of the target, but don't have the target array yet. In that case, numpy/dask has no chance to figure out the shape on its own and you can use `broadcast_to` to give that information explicitly.

More information on broadcasting can be found in [the numpy docu (also note the links at the top)](https://numpy.org/doc/stable/user/basics.broadcasting.html).

Does that make sense?

valeriupredoi commented:
@ledm dis still needed?

valeriupredoi commented:
cheers @ledm - I be nuking this then :bomb:
